### PR TITLE
Add SO3 rotate and unrotate

### PR DESCRIPTION
**Pull request by fantaosha**
## Types of changes

- [ ] Docs change / refactoring / dependency upgrade
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist

- [ ] My code follows the code style of this project.
- [ ] My change requires a change to the documentation.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the **CONTRIBUTING** document.
- [ ] I have completed my CLA (see **CONTRIBUTING**)
- [ ] I have added tests to cover my changes.
- [ ] All new and existing tests passed.

### CODE REVIEW

**Comment by fantaosha** [comment]
I installed [precomit](https://pre-commit.com) and passed on my laptop. However, it looks like I still got errors `ci/circleci: py37_linux` and `ci/circleci: py38_linux` about the formats of python files. Do you know why @luisenp? Thanks!

Screen Shot 2022-01-31 at 10 24 48 AM

**Comment by fantaosha** [comment]
SO3 is almost done. I will start working on SE3 if no new features are requested.

File: theseus/geometry/tests/test_so3.py [hunk 1]
```diff
@@ -148,3 +149,52 @@ def test_adjoint():
...
+        check_compose(so3_1, so3_2)
+
+
+def test_rotate_and_unrotate():
+    rng = torch.Generator()
+    rng.manual_seed(0)
+    for _ in range(10):  # repeat a few times
+        for batch_size in [1, 20, 100]:
+            so3 = _create_random_so3(batch_size, rng)
+            # Tests that rotate works from tensor. unrotate() would work similarly), but
```
Comment (luisenp) [inline theseus/geometry/tests/test_so3.py:169]: This comment can be removed, correct?

File: theseus/geometry/so3.py [hunk 2]
```diff
@@ -12,6 +7,11 @@
 from .lie_group import LieGroup
 from .point_types import Point3
 
+# Copyright (c) Meta Platforms, Inc. and affiliates.
```
Comment (luisenp) [inline theseus/geometry/so3.py:10]: This should be at the top of the file.

**Review by luisenp — APPROVED** [review]
LGTM. Only two minor comments.

**Comment by luisenp** [comment]
For the lint, I think we need to update the precommit config, due to an update in black version. I'll push a fix to main and you should rebase this on top.

**Comment by luisenp** [comment]
I pushed a fix to the precommit tests. @fantaosha can you check if rebasing on top of `main` resolves the CI issues?

File: theseus/geometry/so3.py (hunk 2 above)
Comment (fantaosha) [reply theseus/geometry/so3.py]: This issue is fixed in the newest commit.

File: theseus/geometry/tests/test_so3.py (hunk 1 above)
Comment (fantaosha) [reply theseus/geometry/tests/test_so3.py]: Fixed in the newest commit.

File: theseus/geometry/so3.py [hunk 3]
```diff
@@ -274,3 +277,57 @@ def _copy_impl(self, new_name: Optional[str] = None) -> "SO3":
...
+        self,
+        point: Union[Point3, torch.Tensor],
+        jacobians: Optional[List[torch.Tensor]] = None,
+    ) -> Point3:
+        self._rotate_shape_check(point)
+        batch_size = max(self.shape[0], point.shape[0])
+        if isinstance(point, torch.Tensor):
+            p = point.view(-1, 3, 1)
+        else:
+            p = point.data.view(-1, 3, 1)
```
Comment (mhmukadam) [inline theseus/geometry/so3.py:291]: Do you need a check here to make sure `point` is of the correct size if passed as a `torch.Tensor`?

**Review by mhmukadam — APPROVED** [review]

File: theseus/geometry/so3.py (hunk 3 above)
Comment (fantaosha) [reply theseus/geometry/so3.py:291]: > Do you need a check here to make sure `point` is of the correct size if passed as a `torch.Tensor`? Same for unrotate.

I don't think so. `Union[Point3, torch.Tensor]` has already guaranteed `point` is either `Point3` or `torch.Tensor`.

Comment (luisenp) [reply theseus/geometry/so3.py:291]: I think @mhmukadam meant that when the tensor case is hit (line 289), check that `point.shape` is consistent with what's expected. This is already done in line 286 with the call to `self._rotate_shape_check(point)`.

BTW, note that typing hints are not enforced by Python, it's mostly a help for developers. If the user is not using `mypy`, they won't see any issue with passing arbitrary objects here, until the code crashes somewhere.

**Pull request merged**